### ISSUE TITLE
storage: `MVCCFindSplitKey` performance regression

### ISSUE THREAD

**Issue by erikgrinaker** [cockroachdb/cockroach#89171]
As seen in #88723, `MVCCFindSplitKey_Pebble` has a ~30% performance regression since 22.1.

**Comment by erikgrinaker** [comment]
A CPU profile diff indicates that either `cmpbody` is slower now, or we're doing more comparisons:

```
      flat  flat%   sum%        cum   cum%
     150ms  7.04%  7.04%      150ms  7.04%  cmpbody
     120ms  5.63% 12.68%      410ms 19.25%  github.com/cockroachdb/cockroach/pkg/storage.isValidSplitKey
    -100ms  4.69%  7.98%      -80ms  3.76%  github.com/cockroachdb/pebble.(*Iterator).findNextEntry
      90ms  4.23% 12.21%      290ms 13.62%  github.com/cockroachdb/cockroach/pkg/roachpb.Span.ProperlyContainsKey
     -90ms  4.23%  7.98%     -240ms 11.27%  github.com/cockroachdb/pebble.(*mergingIter).nextEntry
     -80ms  3.76%  4.23%     -150ms  7.04%  github.com/cockroachdb/cockroach/pkg/storage.findSplitKeyUsingIterator
      80ms  3.76%  7.98%     -140ms  6.57%  github.com/cockroachdb/pebble.(*Iterator).nextUserKey
     -70ms  3.29%  4.69%     -360ms 16.90%  github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).Next
     -70ms  3.29%  1.41%     -140ms  6.57%  github.com/cockroachdb/pebble/sstable.(*singleLevelIterator).Next
     -60ms  2.82%  1.41%      -60ms  2.82%  github.com/cockroachdb/pebble.(*Iterator).maybeSampleRead
```

However, I also found a buggy optimization -- the fix in #89172 more than makes up for the regression.

Compares to `master`:

```
name                                     old time/op   new time/op   delta
MVCCFindSplitKey_Pebble/valueSize=32-24    105ms ± 4%     73ms ± 3%  -30.48%  (p=0.000 n=9+10)
```

Compared to 22.1:

```
name                                     old time/op   new time/op   delta
MVCCFindSplitKey_Pebble/valueSize=32-24   88.0ms ± 2%   72.7ms ± 3%  -17.35%  (p=0.000 n=10+10)
```

Even so, I'm going to dig a bit further to figure out where the initial regression came from.

**Comment by erikgrinaker** [comment]
>I'm going to dig a bit further to figure out where the initial regression came from.

It's not Go 1.19, since Go 1.18 shows a similar regression compared to 22.1.

**Comment by erikgrinaker** [comment]
I found the original regression. In `release-22.1`, `keys.NoSplitSpans` is:

https://github.com/cockroachdb/cockroach/blob/13880c9e8d6abc5c7fbd624254f088432b500004/pkg/keys/spans.go#L54

On `master`, it is:

https://github.com/cockroachdb/cockroach/blob/13880c9e8d6abc5c7fbd624254f088432b500004/pkg/keys/spans.go#L55

The keys used by the benchmark, `keys-xxxx`, all sort before `SystemConfigSpan.EndKey`. This means that it avoids the bug in #89172, and the truncated `noSplitSpans` used by `IsValidSplitKey` has a single member, `SystemConfigSpan`. In `master`, however, `keys-xxxx` sorts after all members of `keys.NoSplitSpans`, so it hits the bug in #89172 and has to compare against the entire three-member list of `keys.NoSplitSpans`. This issue is therefore fully resolved by #89172.
